src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,334 @@
+  simp only [←subtype.val_eq_coe, left_coset_assoc],
+end
+| ∞ := rfl
+end
+
+lemma one_smul (x : X') :
+  (1 : B) • x = x :=
+match x with
+| from_coset y := begin
+  change from_coset _ = from_coset _,
+  simp only [←subtype.val_eq_coe, one_left_coset, subtype.ext_iff_val],
+end
+| ∞ := rfl
+end
+
+lemma from_coset_eq_of_mem_range {b : B} (hb : b ∈ f.range) :
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ =
+  from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+begin
+  congr,
+  change b *l f.range = f.range,
+  nth_rewrite 1 [show (f.range : set B) = 1 *l f.range, from (one_left_coset _).symm],
+  rw [left_coset_eq_iff, mul_one],
+  exact subgroup.inv_mem _ hb,
+end
+
+lemma from_coset_ne_of_nin_range {b : B} (hb : b ∉ f.range) :
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ ≠
+  from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+begin
+  intros r,
+  simp only [subtype.mk_eq_mk] at r,
+  change b *l f.range = f.range at r,
+  nth_rewrite 1 [show (f.range : set B) = 1 *l f.range, from (one_left_coset _).symm] at r,
+  rw [left_coset_eq_iff, mul_one] at r,
+  exact hb (inv_inv b ▸ (subgroup.inv_mem _ r)),
+end
+
+
+noncomputable instance : decidable_eq X' := classical.dec_eq _
+
+/--
+Let `τ` be the permutation on `X'` exchanging `f.range` and the point at infinity.
+-/
+noncomputable def tau : SX' := equiv.swap
+(from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) ∞
+
+local notation `τ` := tau f
+
+lemma τ_apply_infinity :
+  τ ∞ = from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+equiv.swap_apply_right _ _
+
+lemma τ_apply_from_coset :
+  τ (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) = ∞ :=
+equiv.swap_apply_left _ _
+
+lemma τ_apply_from_coset' (x : B) (hx : x ∈ f.range) :
+  τ (from_coset ⟨x *l f.range.carrier, ⟨x, rfl⟩⟩) = ∞ :=
+(from_coset_eq_of_mem_range _ hx).symm ▸ τ_apply_from_coset _
+
+lemma τ_symm_apply_from_coset :
+  (equiv.symm τ) (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) = ∞ :=
+by rw [tau, equiv.symm_swap, equiv.swap_apply_left]
+
+lemma τ_symm_apply_infinity :
+  (equiv.symm τ) ∞ = from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+by rw [tau, equiv.symm_swap, equiv.swap_apply_right]
+
+/--
+Let `g : B ⟶ S(X')` be defined as as such that, for any `β : B`, `g(β)` is the function sending
+point at infinity to point at infinity and sending coset `y` to `β *l y`.
+-/
+def G : B →* SX' :=
+{ to_fun := λ β,
+  { to_fun := λ x, β • x,
+    inv_fun := λ x, β⁻¹ • x,
+    left_inv := λ x, by dsimp only; rw [←mul_smul, mul_left_inv, one_smul],
+    right_inv := λ x, by dsimp only; rw [←mul_smul, mul_right_inv, one_smul] },
+  map_one' := by ext; simp [one_smul],
+  map_mul' := λ b1 b2, by ext; simp [mul_smul] }
+
+local notation `g` := G f
+
+/--
+Define `h : B ⟶ S(X')` to be `τ g τ⁻¹`
+-/
+noncomputable def H : B →* SX':=
+{ to_fun := λ β, ((τ).symm.trans (g β)).trans τ,
+  map_one' := by ext; simp,
+  map_mul' := λ b1 b2, by ext; simp }
+
+local notation `h` := H f
+
+/--
```

Comment:
```suggestion
/-!
```
This way, the comment is rendered as a section docstring rather than a docstring for the lemma `g_apply_from_coset`.